api/endpoints.go

```diff
@@ -0,0 +1,302 @@
+	apiEndpoint, _ := getAPIContext(r)
+	if apiEndpoint != nil {
+		return apiEndpoint.Read
+	}
+	return NotFound
+}
+
+// WritePermission returns the write permission for the handler.
+func (eh *endpointHandler) WritePermission(r *http.Request) Permission {
+	apiEndpoint, _ := getAPIContext(r)
+	if apiEndpoint != nil {
+		return apiEndpoint.Write
+	}
+	return NotFound
+}
+
+// ServeHTTP handles the http request.
+func (eh *endpointHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	apiEndpoint, apiRequest := getAPIContext(r)
+	if apiEndpoint == nil || apiRequest == nil {
+		http.NotFound(w, r)
+		return
+	}
+
+	switch r.Method {
+	case http.MethodHead:
+		http.Error(w, "", http.StatusOK)
+		return
+	case http.MethodPost, http.MethodPut:
+		// Read body data.
+		inputData, ok := readBody(w, r)
+		if !ok {
+			return
+		}
+		apiRequest.InputData = inputData
+	case http.MethodGet:
+		// Nothing special to do here.
+	default:
+		http.Error(w, "Unsupported method for the actions API.", http.StatusMethodNotAllowed)
+		return
+	}
+
+	// Execute action function and get response data
+	var responseData []byte
+	var err error
+
+	switch {
+	case apiEndpoint.ActionFn != nil:
+		var msg string
+		msg, err = apiEndpoint.ActionFn(apiRequest)
+		if err == nil {
+			responseData = []byte(msg)
+		}
+
+	case apiEndpoint.DataFn != nil:
+		responseData, err = apiEndpoint.DataFn(apiRequest)
+
+	case apiEndpoint.StructFn != nil:
+		var v interface{}
+		v, err = apiEndpoint.StructFn(apiRequest)
+		if err == nil && v != nil {
+			responseData, err = json.Marshal(v)
+		}
+
+	case apiEndpoint.RecordFn != nil:
+		var rec record.Record
+		rec, err = apiEndpoint.RecordFn(apiRequest)
+		if err == nil && r != nil {
+			responseData, err = marshalRecord(rec, false)
+		}
+
+	case apiEndpoint.HandlerFn != nil:
+		apiEndpoint.HandlerFn(w, r)
+		return
+
+	default:
+		http.Error(w, "Internal server error: Missing handler.", http.StatusInternalServerError)
+		return
+	}
+
+	// Check for handler error.
+	if err != nil {
+		http.Error(w, "Internal server error: "+err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	// Write response.
+	w.Header().Set("Content-Type", apiEndpoint.MimeType+"; charset=utf-8")
+	w.Header().Set("Content-Length", strconv.Itoa(len(responseData)))
+	w.WriteHeader(http.StatusOK)
+	_, err = w.Write(responseData)
+	if err != nil {
+		log.Tracer(r.Context()).Warningf("api: failed to write response: %s", err)
+	}
+}
+
+func readBody(w http.ResponseWriter, r *http.Request) (inputData []byte, ok bool) {
+	// Check for too long content in order to prevent death.
+	if r.ContentLength > 20000000 { // 20MB
+		http.Error(w, "Too much input data.", http.StatusBadRequest)
+		return nil, false
+	}
+
+	// Read and close body.
+	inputData, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		http.Error(w, "Failed to read body: "+err.Error(), http.StatusInternalServerError)
+		return nil, false
+	}
+	r.Body.Close()
```

Comment:
A http.Handler does not need to close the body so this line can be removed (the server closes the body on it's own)